diff -u udt4/src//api.cpp udt4.mod/src//api.cpp
--- udt4/src//api.cpp	2011-08-29 05:16:38.000000000 -0400
+++ udt4.mod/src//api.cpp	2011-09-04 11:46:14.980272272 -0400
@@ -178,6 +178,18 @@
    delete m_pCache;
 }
 
+void CUDTUnited::set_ext_ptr(const UDTSOCKET u, void *ext_ptr)
+{
+       CUDTSocket* s = locate(u);
+       s->ext_ptr = ext_ptr;
+}
+
+void *CUDTUnited::get_ext_ptr(const UDTSOCKET u)
+{
+       CUDTSocket* s = locate(u);
+       return s->ext_ptr;      
+}
+
 int CUDTUnited::startup()
 {
    CGuard gcinit(m_InitLock);
@@ -1538,6 +1550,16 @@
 
 ////////////////////////////////////////////////////////////////////////////////
 
+void CUDT::set_ext_ptr(UDTSOCKET u, void *ext_ptr)
+{
+       return s_UDTUnited.set_ext_ptr(u, ext_ptr);
+}
+
+void *CUDT::get_ext_ptr(UDTSOCKET u)
+{
+       return s_UDTUnited.get_ext_ptr(u);
+}
+
 int CUDT::startup()
 {
    return s_UDTUnited.startup();
@@ -2147,6 +2169,16 @@
 namespace UDT
 {
 
+void set_ext_ptr(UDTSOCKET u, void *ext_ptr)
+{
+       return CUDT::set_ext_ptr(u, ext_ptr);
+}
+
+void *get_ext_ptr(UDTSOCKET u)
+{
+       return CUDT::get_ext_ptr(u);
+}
+
 int startup()
 {
    return CUDT::startup();
diff -u udt4/src//api.h udt4.mod/src//api.h
--- udt4/src//api.h	2011-08-29 05:16:38.000000000 -0400
+++ udt4.mod/src//api.h	2011-09-04 11:47:41.592701760 -0400
@@ -83,6 +83,7 @@
    unsigned int m_uiBackLog;                 // maximum number of connections in queue
 
    int m_iMuxID;                             // multiplexer ID
+   void *ext_ptr;
 
    pthread_mutex_t m_ControlLock;            // lock this socket exclusively for control APIs: bind/listen/connect
 
@@ -104,6 +105,9 @@
 
 public:
 
+    void set_ext_ptr(const UDTSOCKET u, void *ext_ptr);
+    void *get_ext_ptr(const UDTSOCKET u);
+
       // Functionality:
       //    initialize the UDT library.
       // Parameters:
diff -u udt4/src//core.h udt4.mod/src//core.h
--- udt4/src//core.h	2011-08-29 05:16:38.000000000 -0400
+++ udt4.mod/src//core.h	2011-09-04 11:48:36.356973324 -0400
@@ -76,6 +76,8 @@
    ~CUDT();
 
 public: //API
+   static void set_ext_ptr(UDTSOCKET u, void *ext_ptr);
+   static void *get_ext_ptr(UDTSOCKET u);
    static int startup();
    static int cleanup();
    static UDTSOCKET socket(int af, int type = SOCK_STREAM, int protocol = 0);
diff -u udt4/src//udt.h udt4.mod/src//udt.h
--- udt4/src//udt.h	2011-08-29 05:16:38.000000000 -0400
+++ udt4.mod/src//udt.h	2011-09-04 11:49:16.833174033 -0400
@@ -300,6 +300,8 @@
 #undef ERROR
 UDT_API extern const int ERROR;
 
+UDT_API void set_ext_ptr(const UDTSOCKET u, void *ext_ptr);
+UDT_API void *get_ext_ptr(const UDTSOCKET u); 
 UDT_API int startup();
 UDT_API int cleanup();
 UDT_API UDTSOCKET socket(int af, int type, int protocol);
